client/iotcore/client.go

```diff
@@ -186,40 +186,40 @@ func (c *client) UpsertDevice(ctx context.Context,
 				_, err = svc.UpdateCertificate(ctx, paramsUpdateCertificate)
 			}
 		}
-
 		return awsDevice, err
-	} else if err == ErrDeviceNotFound {
-		err = nil
-	}
-
-	var resp *iot.CreateThingOutput
-	if err == nil {
-		resp, err = svc.CreateThing(ctx,
-			&iot.CreateThingInput{
-				ThingName: aws.String(deviceID),
-			})
+	} else if !errors.Is(err, ErrDeviceNotFound) {
+		return nil, fmt.Errorf("unexpected error getting the device: %w", err)
 	}
 
 	var privKey *ecdsa.PrivateKey
-	if err == nil {
-		privKey, err = crypto.NewPrivateKey()
+	privKey, err = crypto.NewPrivateKey()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate key for device: %w", err)
 	}
 
 	var csr []byte
-	if err == nil {
-		csr, err = crypto.NewCertificateSigningRequest(deviceID, privKey)
+	csr, err = crypto.NewCertificateSigningRequest(deviceID, privKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating certificate signing request: %w", err)
+	}
+
+	var resp *iot.CreateThingOutput
+	resp, err = svc.CreateThing(ctx,
```

Comment:
I do not know enough to understand why we add this here?

Reply:
I moved this one down until after we've generated all server side resources before making API calls to AWS.